### PR TITLE
Create example for SSD detection.

### PR DESCRIPTION
**Pull request by hzxie**
Please merge PR #994, #995 and #996.
**Then you can compile this example.**

This example has the same output as [this PyTorch version](https://github.com/amdegroot/ssd.pytorch).
It also indicates that PR #994, #995 and #996 are correct.

I will create README later along with some Python scripts to test this example.

### CODE REVIEW

**Comment by hzxie** [comment]
The following script is used to get the output of PyTorch.
Please note that the output of the image reader in TinyDNN and OpenCV are different, so the final output is different.
Once PyTorch and this example read an image from a text file containing RGB values, they will get the same result.

```
import cv2
import numpy as np
import torch
import math
import itertools

TEST_IMG_PATH = 'Test.jpg'
MODEL_PATH = 'ssd300_mAP_77.43_v2.pth'

def iou(box1, box2):
    '''Compute the intersection over union of two set of boxes, each box is [x1,y1,x2,y2].
    Args:
      box1: (tensor) bounding boxes, sized [N,4].
      box2: (tensor) bounding boxes, sized [M,4].
    Return:
      (tensor) iou, sized [N,M].
    '''
    N = box1.size(0)
    M = box2.size(0)
    lt = torch.max(
        box1[:,:2].unsqueeze(1).expand(N,M,2),  # [N,2] -> [N,1,2] -> [N,M,2]
        box2[:,:2].unsqueeze(0).expand(N,M,2),  # [M,2] -> [1,M,2] -> [N,M,2]
    )
    rb = torch.min(
        box1[:,2:].unsqueeze(1).expand(N,M,2),  # [N,2] -> [N,1,2] -> [N,M,2]
        box2[:,2:].unsqueeze(0).expand(N,M,2),  # [M,2] -> [1,M,2] -> [N,M,2]
    )
    wh = rb - lt  # [N,M,2]
    wh[wh<0] = 0  # clip at 0
    inter = wh[:,:,0] * wh[:,:,1]  # [N,M]
    area1 = (box1[:,2]-box1[:,0]) * (box1[:,3]-box1[:,1])  # [N,]
    area2 = (box2[:,2]-box2[:,0]) * (box2[:,3]-box2[:,1])  # [M,]
    area1 = area1.unsqueeze(1).expand_as(inter)  # [N,] -> [N,1] -> [N,M]
    area2 = area2.unsqueeze(0).expand_as(inter)  # [M,] -> [1,M] -> [N,M]
    iou = inter / (area1 + area2 - inter)
    return iou

def nms(bboxes, scores, threshold=0.5, mode='union'):
    '''Non maximum suppression.
    Args:
      bboxes: (tensor) bounding boxes, sized [N,4].
      scores: (tensor) bbox scores, sized [N,].
      threshold: (float) overlap threshold.
      mode: (str) 'union' or 'min'.
    Returns:
      keep: (tensor) selected indices.
    Ref:
      https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/nms/py_cpu_nms.py
    '''
    x1 = bboxes[:,0]
    y1 = bboxes[:,1]
    x2 = bboxes[:,2]
    y2 = bboxes[:,3]
    areas = (x2-x1) * (y2-y1)
    _, order = scores.sort(0, descending=True)
    print(order)
    keep = []
    while order.numel() > 0:
        i = order[0]
        keep.append(i)
        if order.numel() == 1:
            break
        xx1 = x1[order[1:]].clamp(min=x1[i].item())
        yy1 = y1[order[1:]].clamp(min=y1[i].item())
        xx2 = x2[order[1:]].clamp(max=x2[i].item())
        yy2 = y2[order[1:]].clamp(max=y2[i].item())
        w = (xx2-xx1).clamp(min=0)
        h = (yy2-yy1).clamp(min=0)
        inter = w*h
        if mode == 'union':
            ovr = inter / (areas[i] + areas[order[1:]] - inter)
        elif mode == 'min':
            ovr = inter / areas[order[1:]].clamp(max=areas[i])
        else:
            raise TypeError('Unknown nms mode: %s.' % mode)
        ids = (ovr<=threshold).nonzero().squeeze()
        if ids.numel() == 0:
            break
        order = order[ids+1]
    
    return torch.LongTensor(keep)

ckpt = torch.load(MODEL_PATH)
img = cv2.imread(TEST_IMG_PATH)
img = cv2.resize(img, (300, 300))
img = img[:, :, (2, 1, 0)].transpose((2, 0, 1)).reshape(1, 3, 300, 300)
print(img.shape)
img = torch.from_numpy(img).float().cuda()

conv1_feature = torch.nn.functional.conv2d(img, ckpt['vgg.0.weight'], ckpt['vgg.0.bias'], padding=1)
relu1_feature = torch.nn.functional.relu(conv1_feature)
conv2_feature = torch.nn.functional.conv2d(relu1_feature, ckpt['vgg.2.weight'], ckpt['vgg.2.bias'], padding=1)
relu2_feature = torch.nn.functional.relu(conv2_feature)
pool2_feature = torch.nn.functional.max_pool2d(relu2_feature, kernel_size=2)
conv3_feature = torch.nn.functional.conv2d(pool2_feature, ckpt['vgg.5.weight'], ckpt['vgg.5.bias'], padding=1)
relu3_feature = torch.nn.functional.relu(conv3_feature)
conv4_feature = torch.nn.functional.conv2d(relu3_feature, ckpt['vgg.7.weight'], ckpt['vgg.7.bias'], padding=1)
relu4_feature = torch.nn.functional.relu(conv4_feature)
pool4_feature = torch.nn.functional.max_pool2d(relu4_feature, kernel_size=2)
conv5_feature = torch.nn.functional.conv2d(pool4_feature, ckpt['vgg.10.weight'], ckpt['vgg.10.bias'], padding=1)
relu5_feature = torch.nn.functional.relu(conv5_feature)
conv6_feature = torch.nn.functional.conv2d(relu5_feature, ckpt['vgg.12.weight'], ckpt['vgg.12.bias'], padding=1)
relu6_feature = torch.nn.functional.relu(conv6_feature)
conv7_feature = torch.nn.functional.conv2d(relu6_feature, ckpt['vgg.14.weight'], ckpt['vgg.14.bias'], padding=1)
relu7_feature = torch.nn.functional.relu(conv7_feature)
pool7_feature = torch.nn.functional.max_pool2d(relu7_feature, kernel_size=2, ceil_mode=True)
conv8_feature = torch.nn.functional.conv2d(pool7_feature, ckpt['vgg.17.weight'], ckpt['vgg.17.bias'], padding=1)
relu8_feature = torch.nn.functional.relu(conv8_feature)
conv9_feature = torch.nn.functional.conv2d(relu8_feature, ckpt['vgg.19.weight'], ckpt['vgg.19.bias'], padding=1)
relu9_feature = torch.nn.functional.relu(conv9_feature)
conv10_feature = torch.nn.functional.conv2d(relu9_feature, ckpt['vgg.21.weight'], ckpt['vgg.21.bias'], padding=1)
relu10_feature = torch.nn.functional.relu(conv10_feature)
x = relu10_feature
l2norm_feature = 20 * x * x.pow(2).sum(1).clamp(min=1e-12).rsqrt().expand_as(x)
pool10_feature = torch.nn.functional.max_pool2d(l2norm_feature, kernel_size=2)
conv11_feature = torch.nn.functional.conv2d(pool10_feature, ckpt['vgg.24.weight'], ckpt['vgg.24.bias'], padding=1)
relu11_feature = torch.nn.functional.relu(conv11_feature)
conv12_feature = torch.nn.functional.conv2d(relu11_feature, ckpt['vgg.26.weight'], ckpt['vgg.26.bias'], padding=1)
relu12_feature = torch.nn.functional.relu(conv12_feature)
conv13_feature = torch.nn.functional.conv2d(relu12_feature, ckpt['vgg.28.weight'], ckpt['vgg.28.bias'], padding=1)
relu13_feature = torch.nn.functional.relu(conv13_feature)
pool13_feature = torch.nn.functional.max_pool2d(relu13_feature, kernel_size=3, stride=1, padding=1)
conv14_feature = torch.nn.functional.conv2d(pool13_feature, ckpt['vgg.31.weight'], ckpt['vgg.31.bias'], padding=6, dilation=6)
relu14_feature = torch.nn.functional.relu(conv14_feature)
conv15_feature = torch.nn.functional.conv2d(relu14_feature, ckpt['vgg.33.weight'], ckpt['vgg.33.bias'])
relu15_feature = torch.nn.functional.relu(conv15_feature)

conv16_feature = torch.nn.functional.conv2d(relu15_feature, ckpt['extras.0.weight'], ckpt['extras.0.bias'])
relu16_feature = torch.nn.functional.relu(conv16_feature)
conv17_feature = torch.nn.functional.conv2d(relu16_feature, ckpt['extras.1.weight'], ckpt['extras.1.bias'], stride=2, padding=1)
relu17_feature = torch.nn.functional.relu(conv17_feature)

conv18_feature = torch.nn.functional.conv2d(relu17_feature, ckpt['extras.2.weight'], ckpt['extras.2.bias'])
relu18_feature = torch.nn.functional.relu(conv18_feature)
conv19_feature = torch.nn.functional.conv2d(relu18_feature, ckpt['extras.3.weight'], ckpt['extras.3.bias'], stride=2, padding=1)
relu19_feature = torch.nn.functional.relu(conv19_feature)

conv20_feature = torch.nn.functional.conv2d(relu19_feature, ckpt['extras.4.weight'], ckpt['extras.4.bias'])
relu20_feature = torch.nn.functional.relu(conv20_feature)
conv21_feature = torch.nn.functional.conv2d(relu20_feature, ckpt['extras.5.weight'], ckpt['extras.5.bias'])
relu21_feature = torch.nn.functional.relu(conv21_feature)

conv22_feature = torch.nn.functional.conv2d(relu21_feature, ckpt['extras.6.weight'], ckpt['extras.6.bias'])
relu22_feature = torch.nn.functional.relu(conv22_feature)
conv23_feature = torch.nn.functional.conv2d(relu22_feature, ckpt['extras.7.weight'], ckpt['extras.7.bias'])
relu23_feature = torch.nn.functional.relu(conv23_feature)

conv24_feature = torch.nn.functional.conv2d(l2norm_feature, ckpt['loc.0.weight'], ckpt['loc.0.bias'], padding=1)
conv25_feature = torch.nn.functional.conv2d(relu15_feature, ckpt['loc.1.weight'], ckpt['loc.1.bias'], padding=1)
conv26_feature = torch.nn.functional.conv2d(relu17_feature, ckpt['loc.2.weight'], ckpt['loc.2.bias'], padding=1)
conv27_feature = torch.nn.functional.conv2d(relu19_feature, ckpt['loc.3.weight'], ckpt['loc.3.bias'], padding=1)
conv28_feature = torch.nn.functional.conv2d(relu21_feature, ckpt['loc.4.weight'], ckpt['loc.4.bias'], padding=1)
conv29_feature = torch.nn.functional.conv2d(relu23_feature, ckpt['loc.5.weight'], ckpt['loc.5.bias'], padding=1)

conv30_feature = torch.nn.functional.conv2d(l2norm_feature, ckpt['conf.0.weight'], ckpt['conf.0.bias'], padding=1)
conv31_feature = torch.nn.functional.conv2d(relu15_feature, ckpt['conf.1.weight'], ckpt['conf.1.bias'], padding=1)
conv32_feature = torch.nn.functional.conv2d(relu17_feature, ckpt['conf.2.weight'], ckpt['conf.2.bias'], padding=1)
conv33_feature = torch.nn.functional.conv2d(relu19_feature, ckpt['conf.3.weight'], ckpt['conf.3.bias'], padding=1)
conv34_feature = torch.nn.functional.conv2d(relu21_feature, ckpt['conf.4.weight'], ckpt['conf.4.bias'], padding=1)
conv35_feature = torch.nn.functional.conv2d(relu23_feature, ckpt['conf.5.weight'], ckpt['conf.5.bias'], padding=1)

x = [conv24_feature, conv25_feature, conv26_feature, conv27_feature, conv28_feature, conv29_feature]
_x = [i.permute(0, 2, 3, 1).contiguous() for i in x]
locations = torch.cat([l.view(l.size(0), -1) for l in _x], dim=1).reshape(8732, 4)

y = [conv30_feature, conv31_feature, conv32_feature, conv33_feature, conv34_feature, conv35_feature]
_y = [i.permute(0, 2, 3, 1).contiguous() for i in y]
conf = torch.cat([c.view(c.size(0), -1) for c in _y], dim=1).reshape(8732, 21)
conf = torch.nn.functional.softmax(conf, dim=1)
max_conf, labels = conf.max(1)

'''Compute default box sizes with scale and aspect transform.'''
scale = 300.
steps = [s / scale for s in (8, 16, 32, 64, 100, 300)]
sizes = [s / scale for s in (30, 60, 111, 162, 213, 264, 315)]
aspect_ratios = ((2,), (2,3), (2,3), (2,3), (2,), (2,))
feature_map_sizes = (38, 19, 10, 5, 3, 1)
num_layers = len(feature_map_sizes)

boxes = []
for i in range(num_layers):
    fmsize = feature_map_sizes[i]
    for h,w in itertools.product(range(fmsize), repeat=2):
        cx = (w + 0.5)*steps[i]
        cy = (h + 0.5)*steps[i]
        s = sizes[i]
        boxes.append((cx, cy, s, s))
        s = math.sqrt(sizes[i] * sizes[i+1])
        boxes.append((cx, cy, s, s))
        s = sizes[i]
        for ar in aspect_ratios[i]:
            boxes.append((cx, cy, s * math.sqrt(ar), s / math.sqrt(ar)))
            boxes.append((cx, cy, s / math.sqrt(ar), s * math.sqrt(ar)))
default_boxes = torch.Tensor(boxes).cuda()

'''Transform predicted loc/conf back to real bbox locations and class labels.'''
variances = [0.1, 0.2]
boxes = torch.cat((
    default_boxes[:, :2] + locations[:, :2] * variances[0] * default_boxes[:, 2:],
    default_boxes[:, 2:] * torch.exp(locations[:, 2:] * variances[1])), 1)

boxes[:, :2] -= boxes[:, 2:] / 2
boxes[:, 2:] += boxes[:, :2]

keep_idx = nms(boxes[labels.nonzero()].squeeze(1) * 300, max_conf[labels.nonzero()])
print(keep_idx)
```

**Comment by edgarriba** [comment]
@hzxie very cool !! do you have any metrics in terms of accuracy and speed in order to be able to compare results ?

**Comment by edgarriba** [comment]
@hzxie awesome. Would be nice if you can include as much as possible documentation around this comparison in the mentioned readme. I will review asap as possible the code. Great contribution man !

**Comment by beru** [comment]
This library used to use OpenCV with `USE_OPENCV` option but stopped relying on it.
I think it's OK to use it as long as the sample can be built with/without `option USE_OPENCV` in CMakeLists.txt.

**Comment by hzxie** [comment]
@beru 
The code above is used to visualize the detection results.
The TinyDNN version doesn't reply on OpenCV.

**Comment by beru** [comment]
@hzxie I see. What I tride to mention is incorporating the visualization with OpenCV is OK as long as they are taken care of with proper preprocessor macros.

I just installed Anaconda and PyTorch and converting the [model](https://s3.amazonaws.com/amdegroot-models/ssd300_mAP_77.43_v2.pth) introduced in [here](https://github.com/amdegroot/ssd.pytorch) with the script and it looks like it takes quite a while....

**Comment by hzxie** [comment]
Here's the detection result.
screenshot from 2018-10-21 13-56-06

**Comment by beru** [comment]
@hzxie How long does it usually take to convert the model with the script?
And is this the right model file?
https://s3.amazonaws.com/amdegroot-models/ssd300_mAP_77.43_v2.pth

**Comment by beru** [comment]
> Here's the detection result of the following original image

I see. So the network has the ability to fatten airplanes?

**Comment by hzxie** [comment]
@beru
The model convention process needs a long time (I'm testing it how long it will take).
If you can see the outputs from the Python script, just wait patiently.

File: examples/ssd_detection/test.cpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+}
+
+int main(int argc, char** argv) {
+  if (argc != 3) {
+    std::cout << "Usage: example_ssd_test model_folder_path img_file_path";
+    return -1;
+  }
+  std::vector<tiny_dnn::network<tiny_dnn::sequential>> nets;
+
+  construct_nets(nets, argv[1]);
```
Comment (beru) [inline examples/ssd_detection/test.cpp]: User specified `model_folder_path` argument may not contain trailing path separator letter, so adding it would be helpful.

**Comment by beru** [comment]
It seems VOC dataset has 20 classes.
https://github.com/amdegroot/ssd.pytorch/blob/8dd38657a3b1df98df26cf18be9671647905c2a0/data/voc0712.py#L20

Tested with below image.
hqdefault

`example_ssd_test` output.
```
Bounding box coordinates:
x_min = 185.679, x_max = 282.387, y_min = 49.0915, y_max = 255.018, class = 15, score = 0.877027
```

According to @hzxie , SSD network input size is 300 x 300, so I resized the image to the size and checked bounding box coordinates.

image

Obviously, there isn't a horse in the image, so the network detected a person.

**Comment by beru** [comment]
Tested with another image.
highland-cow-and-calf

Here's output from `example_ssd_test`.
```
Bounding box coordinates:
x_min = 10.3258, x_max = 206.018, y_min = 36.8716, y_max = 277.781, class = 10, score = 0.952407
x_min = 204.504, x_max = 286.855, y_min = 145.433, y_max = 274.785, class = 10, score = 0.59865
```
image

**Comment by beru** [comment]
Tested with this woody horse.
young a orchard 610 img_1134_2

`example_ssd_test` output.
```
Bounding box coordinates:
x_min = 48.6329, x_max = 273.264, y_min = 59.2574, y_max = 241.993, class = 15, score = 0.469386
```

image

A person? It doesn't look like anything to me...

**Comment by beru** [comment]
Another try with this image.
horse-suit1

Here's the result.
```
Bounding box coordinates:
x_min = 174.919, x_max = 242.278, y_min = 47.6814, y_max = 265.807, class = 15, score = 0.532353
```
image

**Comment by hzxie** [comment]
There're some missing objects due to SSD is not perfect.
If SSD is perfect, the mAP of SSD should be 100% instead of 78%.
LGTM.
